### PR TITLE
DisplayInput Component

### LINKED ISSUE #245: New Component / Display Input
This is a special type of input and should be considered a unique component. While the input in issue #244 is primarily for forms and creation flows with multiple inputs. This component is more for to display a list of details but allows for inline editing. In this case the user may only need to edit one (if any) of the details and it would not make sense to display info as a form. 

Some MX application examples of this pattern are: account details, transaction details, goal details, etc.

The spec can be found in Aggro style guide > Form Elements > Display Input: https://app.frontify.com/d/KHQZnlj1XMIO/web-patterns

Example: 
image

### PR DESCRIPTION
**Pull request by iheartrachie**
Creates the `DisplayInput` component and closes #245. The component is used for text input and is fully responsive. Its functionality should be similar to the `SimpleInput` component.

## Example
```
_handleInputShowHint () {
  this.setState({
    showHint: true
  });
},

 _handleInputHideHint () {
  this.setState({
    showHint: false
  });
},

_handleInputFocus () {
  this.setState({
    statusMessage: null,
    showHint: false
  });
},

_handleInputStatusMessage () {
  this.setState({
    statusMessage: {
      type: 'success',
      message: 'Saved!'
    },
    valid: true
  });
},

<DisplayInput
  hint='Click to Edit'
  label='Display Input'
  onBlur={this._handleInputStatusMessage}
  onFocus={this._handleInputFocus}
  onMouseOut={this._handleInputHideHint}
  onMouseOver={this._handleInputShowHint}
  placeholder='Type something'
  showHint={this.state.showHint}
  status={this.state.statusMessage}
  valid={true}
/>
```
**Successful input**
7ypatz535p

**Error on input**
lxv4tyrkz8

## Props
`hint` _string_ **optional**
Hint text to display to user on input hover

`label` _string_ **optional**
Input label shown to the left of the input field

`placeholder` _string_ **optional**
Placeholder text shown inside the input field

`primaryColor` _string_ **optional**
Default: PRIMARY
A string that changes the brand color if there is one. Used on input focus and for the hint color

`showHint` _boolean_ **optional**
A boolean that controls when the hint text is displayed.

`status` _Object_ **optional**
An Object that contains a status type (error/success) and a status message that is displayed to the right of the input field.

`valid` _boolean_ **optional**
Default: true
A boolean that marks the field as valid or not.

### CODE REVIEW

**Comment by iheartrachie** [comment]
I'm a little unsure about how I'm doing the event handlers. They feel cumbersome to me. Any suggestions for alternatives would be greatly appreciated! :heart:

**Comment by lukeschunk** [comment]
I could be wrong about this - but don't we want the mouse/click/etc handlers to be at the component level and then bubbled up to the top level via props? @cerinman ?

File: demo/app.js [hunk 1]
```diff
@@ -628,6 +629,35 @@ const Demo = React.createClass({
...
+    this.setState({
+      showHint: true
+    });
+  },
+
+  _handleInputHideHint () {
+    this.setState({
+      showHint: false
+    });
+  },
```
Comment (cerinman) [inline demo/app.js]: You can combine these two functions.

```
_toggleShowInputHint () {
  this.setstate({
    showHint: !this.state.showHint
  });
}
```

**Comment by paniclater** [comment]
@lukeschunk by passing in these functions to the component we get better reusability and the benefit of having the input be able to interact with other portions of the app without having any knowledge of them.

File: demo/app.js [hunk 2]
```diff
@@ -961,6 +991,18 @@ const Demo = React.createClass({
         />
         <br/><br/>
         <SearchInput />
+        <DisplayInput
+          hint='Click to Edit'
+          label='Display Input'
+          onBlur={this._handleInputStatusMessage}
+          onFocus={this._handleInputFocus}
+          onMouseOut={this._handleInputHideHint}
+          onMouseOver={this._handleInputShowHint}
```
Comment (cerinman) [inline demo/app.js]: Ah, hadn't seen that you had `onMouseOut` and `onMouseOver` listeners added.  My previous comment about combing the two functions might not work depending on how you have all that setup.

**Comment by lukeschunk** [comment]
@paniclater my idea wouldn't change the reusability - it would be like how tumen handled his recent SearchInput component. Put the click/blur handlers as props, and then put the functions on the top level. It just seems to be the pattern we usually follow - I'm not opposed to this though, just haven't seen it before.

File: src/components/DisplayInput.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      primaryColor: StyleConstants.Colors.PRIMARY,
+      valid: true
+    };
+  },
+
+  render () {
+    const styles = this.styles();
+
+    return (
+      <div style={Object.assign({}, styles.wrapper)}>
```
Comment (cerinman) [inline src/components/DisplayInput.js]: No need for the `Object.assign` if your not merging two or more style objects together.

File: demo/app.js (hunk 2 above)
Comment (iheartrachie) [inline demo/app.js]: This is the best way I could find to handle a hover state. Is there a better way?

**Comment by paniclater** [comment]
@lukeschunk I see what you mean now. I misunderstood what you were trying to say. I think the benefit of passing in the handlers as props the way Tumen did is that you can specify what element of the component is being targeted. If the entire component is going to respond to blur or click then I think passing them in as props is overkill.

**Comment by lukeschunk** [comment]
@paniclater sounds good - that's kind of what I was wondering - the only benefit I can see is if they're props the person using the component will have an explicit reminder of 'hey if i want anything to happen on hover I have to pass a prop' - this way it might be easier to forget, but I agree it is more code and could be overkill.

File: src/components/DisplayInput.js [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+  },
+
+  styles () {
+    return {
+      wrapper: Object.assign({
+        borderBottom: this.props.valid ? '1px solid ' + StyleConstants.Colors.ASH : '1px solid ' + StyleConstants.Colors.STRAWBERRY,
+        borderLeft: '1px solid transparent',
+        borderRadius: 3,
+        borderRight: '1px solid transparent',
+        borderTop: '1px solid transparent',
```
Comment (cerinman) [inline src/components/DisplayInput.js]: I'm not an css expert so some one correct me if I'm wrong here, but specifying the borders for left, right, and top is not required here right?  By omitting them they should end up as none I would think.

File: src/components/DisplayInput.js [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+        transition: 'all .2s ease-in',
+        WebkitAppearance: 'none',
+        whiteSpace: 'nowrap',
+        width: '100%',
+
+        ':focus': {
+          borderBottom: this.props.valid ? '1px solid ' + this.props.primaryColor : '1px solid ' + StyleConstants.Colors.STRAWBERRY,
+          borderLeft: '1px solid transparent',
+          borderRight: '1px solid transparent',
+          borderTop: '1px solid transparent',
```
Comment (cerinman) [inline src/components/DisplayInput.js]: same here

File: demo/app.js (hunk 2 above)
Comment (cerinman) [inline demo/app.js]: @tumentumurchudur do you want to take a look at this PR when you have a sec?  You always see more than I do and maybe you might have an answer around using `onMouseOut` and `onMouseOver` for showing/hiding the hint.

File: src/components/DisplayInput.js (hunk 4 above)
Comment (iheartrachie) [inline src/components/DisplayInput.js]: Yeah you're right. I'll remove them.

File: src/components/DisplayInput.js [hunk 6]
```diff
@@ -0,0 +1,145 @@
...
+      valid: true
+    };
+  },
+
+  render () {
+    const styles = this.styles();
+
+    return (
+      <div style={Object.assign({}, styles.wrapper)}>
+        {this.props.label ?
```
Comment (tumentumurchudur) [inline src/components/DisplayInput.js]: I would put the `divs` in `()`
```javascript
{this.props.label ? (
  <div key='label' style={styles.label}>
    <div style={styles.labelText}>{this.props.label}</div>
  </div>) : null}
```

**Comment by tumentumurchudur** [comment]
@iheartrachie  Can you post some screenshots?

**Comment by tumentumurchudur** [comment]
@iheartrachie never mind. Just saw the gifs.

File: src/components/DisplayInput.js [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+          {...this.props}
+          key='input'
+          label={this.props.label}
+          style={styles.input}
+          type='text'
+        />
+        <div style={styles.hint}>
+          <div style={styles.hintText}>
+            {this.props.showHint && !this.props.status ? <div>{this.props.hint}</div> : null}
+            {this.props.status ? <div style={styles[this.props.status.type]}>{this.props.status.message}</div> : null}
```
Comment (tumentumurchudur) [inline src/components/DisplayInput.js]: Same here. Put `div` in `()`

File: src/components/DisplayInput.js (hunk 6 above)
Comment (iheartrachie) [inline src/components/DisplayInput.js]: Is this because I have JSX inside of JS?

File: src/components/DisplayInput.js [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+    };
+  },
+
+  render () {
+    const styles = this.styles();
+
+    return (
+      <div style={styles.wrapper}>
+        {this.props.label ?
+          <div key='label' style={styles.label}>
```
Comment (tumentumurchudur) [inline src/components/DisplayInput.js]: I don't see a reason to have `key='label'` here. You can remove that.

Comment (iheartrachie) [inline src/components/DisplayInput.js]: Without it I get a `Radium requires each element with interactive styles to have a unique key` error.

Comment (tumentumurchudur) [inline src/components/DisplayInput.js]: Yeah, Radium needs it. 👍

File: demo/app.js (hunk 2 above)
Comment (tumentumurchudur) [inline demo/app.js]: I think it is fine the way @iheartrachie has it. 🍪

File: src/components/DisplayInput.js [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+  render () {
+    const styles = this.styles();
+
+    return (
+      <div style={styles.wrapper}>
+        {this.props.label ? (
+          <div style={styles.label}>
+            <div style={styles.labelText}>{this.props.label}</div>
+          </div>) : null
+        }
```
Comment (tumentumurchudur) [inline src/components/DisplayInput.js]: Sorry for being nitpicky. Let's move `}` up next to `null}`

**Comment by tumentumurchudur** [comment]
This looks awesome. :shipit:

**Comment by cerinman** [comment]
thanks @tumentumurchudur!  🍪  Your reviewing skills are legendary..

**Pull request merged**